File: machine/cpp/machine/devices/audio/sample_transfer.cpp
#include "machine/devices/audio/sample_transfer.h"

#include "machine/bus/io.h"
#include "machine/cpu/cpu.h"
#include "machine/devices/audio/sample_memory.h"
#include "machine/devices/dma/controller.h"
#include "machine/memory/memory.h"
#include "machine/scheduler/budget.h"
#include "machine/scheduler/device.h"

namespace bmsx {

ApuSampleTransfer::ApuSampleTransfer(
	Memory& memory,
	ApuSampleMemory& sampleMemory,
	DmaController& dma,
	DeviceScheduler& scheduler
)
	: m_memory(memory)
	, m_sampleMemory(sampleMemory)
	, m_dma(dma)
	, m_scheduler(scheduler) {}

void ApuSampleTransfer::reset() {
	cancelBatch();
	m_fifo.fill(0u);
	clearFifo();
	m_currentAddress = 0u;
	m_dataLatch = 0u;
	m_mode = APU_TRANSFER_MODE_STOP;
	m_timingCarry = 0;
	m_memory.writeIoValue(IO_APU_TRANSFER_ADDRESS, valueNumber(0.0));
	m_memory.writeIoValue(IO_APU_TRANSFER_DATA, valueNumber(0.0));
	m_memory.writeIoValue(IO_APU_TRANSFER_CONTROL, valueNumber(0.0));
	updateDmaRequests();
}

void ApuSampleTransfer::dispose() {
	cancelBatch();
	m_dma.setApuDmaWriteReady(false);
	m_dma.setApuDmaReadReady(false);
}

void ApuSampleTransfer::setTiming(i64 cpuHz, i64 nowCycles) {
	if (m_cpuHz == cpuHz) {
		return;
	}
	cancelBatch();
	m_cpuHz = cpuHz;
	m_timingCarry = 0;
	scheduleBatch(nowCycles);
}

auto ApuSampleTransfer::statusBits() const -> u32 {
	const bool readRequest = m_mode == APU_TRANSFER_MODE_DMA_READ
		&& m_fifoCount == APU_TRANSFER_FIFO_WORD_CAPACITY;
	const bool writeRequest = m_mode == APU_TRANSFER_MODE_DMA_WRITE && m_fifoCount == 0u;
	u32 status = 0u;
	if (readRequest || writeRequest) {
		status |= APU_STATUS_DMA_REQUEST;
	}
	if (readRequest) {
		status |= APU_STATUS_DMA_READ_REQUEST;
	}
	if (writeRequest) {
		status |= APU_STATUS_DMA_WRITE_REQUEST;
	}
	if (m_scheduledWords != 0u) {
		status |= APU_STATUS_TRANSFER_BUSY;
	}
	return status;
}

auto ApuSampleTransfer::captureState(i64 nowCycles) const -> ApuSampleTransferState {
	ApuSampleTransferState state;
	state.transferAddressWord = m_memory.readIoU32(IO_APU_TRANSFER_ADDRESS);
	state.transferDataWord = m_dataLatch;
	state.transferControlWord = m_memory.readIoU32(IO_APU_TRANSFER_CONTROL);
	state.currentAddress = m_currentAddress;
	state.fifoWords = m_fifo;
	state.fifoReadIndex = m_fifoReadIndex;
	state.fifoWriteIndex = m_fifoWriteIndex;
	state.fifoCount = m_fifoCount;
	state.timingCarry = m_timingCarry;
	state.scheduledWords = m_scheduledWords;
	state.scheduledCycles = m_scheduledWords == 0u ? 0 : m_serviceDeadline - nowCycles;
	return state;
}

void ApuSampleTransfer::restoreState(const ApuSampleTransferState& state, i64 nowCycles) {
	m_scheduler.cancelDeviceService(DEVICE_SERVICE_APU_TRANSFER);
	m_fifo = state.fifoWords;
	m_fifoReadIndex = state.fifoReadIndex;
	m_fifoWriteIndex = state.fifoWriteIndex;
	m_fifoCount = state.fifoCount;
	m_currentAddress = state.currentAddress;
	m_dataLatch = state.transferDataWord;
	m_mode = state.transferControlWord & APU_TRANSFER_MODE_MASK;
	m_timingCarry = state.timingCarry;
	m_scheduledWords = state.scheduledWords;
	m_serviceDeadline = nowCycles + state.scheduledCycles;
	m_memory.writeIoValue(IO_APU_TRANSFER_ADDRESS, valueNumber(static_cast<f64>(state.transferAddressWord)));
	m_memory.writeIoValue(IO_APU_TRANSFER_DATA, valueNumber(static_cast<f64>(state.transferDataWord)));
	m_memory.writeIoValue(IO_APU_TRANSFER_CONTROL, valueNumber(static_cast<f64>(state.transferControlWord)));
	updateDmaRequests();
	if (m_scheduledWords != 0u) {
		m_scheduler.scheduleDeviceService(DEVICE_SERVICE_APU_TRANSFER, m_serviceDeadline);
	}
}

void ApuSampleTransfer::writeAddress(u32 word) {
	m_currentAddress = word & (APU_SAMPLE_RAM_ADDRESS_MASK & ~(IO_WORD_SIZE - 1u));
}

auto ApuSampleTransfer::readCpuData() const -> u32 {
	return m_dataLatch;
}

auto ApuSampleTransfer::readDmaData() -> u32 {
	if (m_fifoCount != 0u) {
		m_dataLatch = popFifo();
		m_memory.writeIoValue(IO_APU_TRANSFER_DATA, valueNumber(static_cast<f64>(m_dataLatch)));
		updateDmaRequests();
		scheduleBatch(m_scheduler.currentNowCycles());
	}
	return m_dataLatch;
}

void ApuSampleTransfer::writeCpuData(u32 word) {
	m_dataLatch = word;
	if (m_mode == APU_TRANSFER_MODE_MANUAL_WRITE) {
		m_sampleMemory.writeWord(m_currentAddress, word);
		m_currentAddress = (m_currentAddress + IO_WORD_SIZE) & APU_SAMPLE_RAM_ADDRESS_MASK;
	}
}

void ApuSampleTransfer::writeDmaData(u32 word) {
	m_dataLatch = word;
	pushFifo(word);
	updateDmaRequests();
	scheduleBatch(m_scheduler.currentNowCycles());
}

void ApuSampleTransfer::writeControl(u32 word) {
	const u32 mode = word & APU_TRANSFER_MODE_MASK;
	if (mode != m_mode) {
		cancelBatch();
		clearFifo();
		m_timingCarry = 0;
		m_mode = mode;
	}
	updateDmaRequests();
	scheduleBatch(m_scheduler.currentNowCycles());
}

void ApuSampleTransfer::completeService() {
	const i64 completedDeadline = m_serviceDeadline;
	m_scheduler.cancelDeviceService(DEVICE_SERVICE_APU_TRANSFER);
	completeBatch();
	scheduleBatch(completedDeadline);
}

void ApuSampleTransfer::clearFifo() {
	m_fifoReadIndex = 0u;
	m_fifoWriteIndex = 0u;
	m_fifoCount = 0u;
}

void ApuSampleTransfer::pushFifo(u32 word) {
	if (m_fifoCount == APU_TRANSFER_FIFO_WORD_CAPACITY) {
		return;
	}
	m_fifo[m_fifoWriteIndex] = word;
	m_fifoWriteIndex = (m_fifoWriteIndex + 1u) & (APU_TRANSFER_FIFO_WORD_CAPACITY - 1u);
	m_fifoCount += 1u;
}

auto ApuSampleTransfer::popFifo() -> u32 {
	const u32 word = m_fifo[m_fifoReadIndex];
	m_fifoReadIndex = (m_fifoReadIndex + 1u) & (APU_TRANSFER_FIFO_WORD_CAPACITY - 1u);
	m_fifoCount -= 1u;
	return word;
}

void ApuSampleTransfer::completeBatch() {
	const u32 transferWords = m_scheduledWords;
	m_scheduledWords = 0u;
	if (m_mode == APU_TRANSFER_MODE_DMA_READ) {
		for (u32 index = 0u; index < transferWords; index += 1u) {
			pushFifo(m_sampleMemory.readWord(m_currentAddress));
			m_currentAddress = (m_currentAddress + IO_WORD_SIZE) & APU_SAMPLE_RAM_ADDRESS_MASK;
		}
	} else {
		for (u32 index = 0u; index < transferWords; index += 1u) {
			m_sampleMemory.writeWord(m_currentAddress, popFifo());
			m_currentAddress = (m_currentAddress + IO_WORD_SIZE) & APU_SAMPLE_RAM_ADDRESS_MASK;
		}
	}
	updateDmaRequests();
}

void ApuSampleTransfer::scheduleBatch(i64 anchorCycle) {
	if (m_scheduledWords != 0u) {
		return;
	}
	u32 transferWords = 0u;
	if (m_mode == APU_TRANSFER_MODE_DMA_WRITE) {
		transferWords = m_fifoCount;
	} else if (m_mode == APU_TRANSFER_MODE_DMA_READ) {
		transferWords = APU_TRANSFER_FIFO_WORD_CAPACITY - m_fifoCount;
	}
	if (transferWords == 0u) {
		m_serviceDeadline = 0;
		return;
	}
	const i64 transferCycles = cyclesUntilBudgetUnits(
		m_cpuHz,
		APU_TRANSFER_WORDS_PER_SECOND,
		m_timingCarry,
		transferWords
	);
	m_timingCarry = (static_cast<i64>(APU_TRANSFER_WORDS_PER_SECOND) * transferCycles + m_timingCarry) % m_cpuHz;
	m_scheduledWords = transferWords;
	m_serviceDeadline = anchorCycle + transferCycles;
	m_scheduler.scheduleDeviceService(DEVICE_SERVICE_APU_TRANSFER, m_serviceDeadline);
}

void ApuSampleTransfer::cancelBatch() {
	m_scheduler.cancelDeviceService(DEVICE_SERVICE_APU_TRANSFER);
	m_scheduledWords = 0u;
	m_serviceDeadline = 0;
}

void ApuSampleTransfer::updateDmaRequests() {
	m_dma.setApuDmaReadReady(m_mode == APU_TRANSFER_MODE_DMA_READ
		&& m_fifoCount == APU_TRANSFER_FIFO_WORD_CAPACITY);
	m_dma.setApuDmaWriteReady(m_mode == APU_TRANSFER_MODE_DMA_WRITE && m_fifoCount == 0u);
}

} // namespace bmsx

File: machine/ts/machine/devices/audio/sample_transfer.ts
import {
	IO_APU_TRANSFER_ADDRESS,
	IO_APU_TRANSFER_CONTROL,
	IO_APU_TRANSFER_DATA,
} from '../../bus/io';
import type { Memory } from '../../memory/memory';
import { cyclesUntilBudgetUnits } from '../../scheduler/budget';
import { DEVICE_SERVICE_APU_TRANSFER, type DeviceScheduler } from '../../scheduler/device';
import type { DmaController } from '../dma/controller';
import {
	APU_STATUS_DMA_READ_REQUEST,
	APU_STATUS_DMA_REQUEST,
	APU_STATUS_DMA_WRITE_REQUEST,
	APU_STATUS_TRANSFER_BUSY,
	APU_SAMPLE_RAM_ADDRESS_MASK,
	APU_TRANSFER_FIFO_WORD_CAPACITY,
	APU_TRANSFER_MODE_DMA_READ,
	APU_TRANSFER_MODE_DMA_WRITE,
	APU_TRANSFER_MODE_MANUAL_WRITE,
	APU_TRANSFER_MODE_MASK,
	APU_TRANSFER_MODE_STOP,
	APU_TRANSFER_WORDS_PER_SECOND,
} from './contracts';
import type { ApuSampleMemory } from './sample_memory';
import type { ApuSampleTransferState } from './save_state';

// This datapath is owned exclusively by ApuServiceClock. It has no MMIO or
// scheduler entry point of its own, so every state transition shares the APU
// clock domain's transfer-before-DAC ordering.
export class ApuSampleTransfer {
	private readonly fifoWords = new Uint32Array(APU_TRANSFER_FIFO_WORD_CAPACITY);
	private fifoReadIndex = 0;
	private fifoWriteIndex = 0;
	private fifoCount = 0;
	private currentAddress = 0;
	private dataLatch = 0;
	private mode = APU_TRANSFER_MODE_STOP;
	private cpuHz = 1;
	private timingCarry = 0;
	scheduledWords = 0;
	scheduledDeadline = 0;

	public constructor(
		private readonly memory: Memory,
		private readonly sampleMemory: ApuSampleMemory,
		private readonly dma: DmaController,
		private readonly scheduler: DeviceScheduler,
	) {}

	public reset(): void {
		this.cancelBatch();
		this.fifoWords.fill(0);
		this.clearFifo();
		this.currentAddress = 0;
		this.dataLatch = 0;
		this.mode = APU_TRANSFER_MODE_STOP;
		this.timingCarry = 0;
		this.memory.writeIoValue(IO_APU_TRANSFER_ADDRESS, 0);
		this.memory.writeIoValue(IO_APU_TRANSFER_DATA, 0);
		this.memory.writeIoValue(IO_APU_TRANSFER_CONTROL, 0);
		this.updateDmaRequests();
	}

	public dispose(): void {
		this.cancelBatch();
		this.dma.setApuDmaWriteReady(false);
		this.dma.setApuDmaReadReady(false);
	}

	public setTiming(cpuHz: number, nowCycles: number): void {
		if (this.cpuHz === cpuHz) {
			return;
		}
		this.cancelBatch();
		this.cpuHz = cpuHz;
		this.timingCarry = 0;
		this.scheduleBatch(nowCycles);
	}

	public statusBits(): number {
		const dmaWriteRequest = this.mode === APU_TRANSFER_MODE_DMA_WRITE && this.fifoCount === 0;
		const dmaReadRequest = this.mode === APU_TRANSFER_MODE_DMA_READ && this.fifoCount === APU_TRANSFER_FIFO_WORD_CAPACITY;
		return ((dmaWriteRequest || dmaReadRequest ? APU_STATUS_DMA_REQUEST : 0)
			| (dmaReadRequest ? APU_STATUS_DMA_READ_REQUEST : 0)
			| (dmaWriteRequest ? APU_STATUS_DMA_WRITE_REQUEST : 0)
			| (this.scheduledWords !== 0 ? APU_STATUS_TRANSFER_BUSY : 0)) >>> 0;
	}

	public captureState(nowCycles: number): ApuSampleTransferState {
		return {
			fifoWords: Array.from(this.fifoWords),
			fifoReadIndex: this.fifoReadIndex,
			fifoWriteIndex: this.fifoWriteIndex,
			fifoCount: this.fifoCount,
			transferAddressWord: this.memory.readIoU32(IO_APU_TRANSFER_ADDRESS),
			transferDataWord: this.dataLatch,
			transferControlWord: this.memory.readIoU32(IO_APU_TRANSFER_CONTROL),
			currentAddress: this.currentAddress,
			timingCarry: this.timingCarry,
			scheduledWords: this.scheduledWords,
			scheduledCycles: this.scheduledWords === 0 ? 0 : this.scheduledDeadline - nowCycles,
		};
	}

	public restoreState(state: ApuSampleTransferState, nowCycles: number): void {
		this.scheduler.cancelDeviceService(DEVICE_SERVICE_APU_TRANSFER);
		for (let index = 0; index < APU_TRANSFER_FIFO_WORD_CAPACITY; index += 1) {
			this.fifoWords[index] = state.fifoWords[index]!;
		}
		this.fifoReadIndex = state.fifoReadIndex;
		this.fifoWriteIndex = state.fifoWriteIndex;
		this.fifoCount = state.fifoCount;
		this.currentAddress = state.currentAddress;
		this.dataLatch = state.transferDataWord;
		this.mode = state.transferControlWord & APU_TRANSFER_MODE_MASK;
		this.timingCarry = state.timingCarry;
		this.scheduledWords = state.scheduledWords;
		this.scheduledDeadline = nowCycles + state.scheduledCycles;
		this.memory.writeIoValue(IO_APU_TRANSFER_ADDRESS, state.transferAddressWord);
		this.memory.writeIoValue(IO_APU_TRANSFER_DATA, state.transferDataWord);
		this.memory.writeIoValue(IO_APU_TRANSFER_CONTROL, state.transferControlWord);
		this.updateDmaRequests();
		if (this.scheduledWords !== 0) {
			this.scheduler.scheduleDeviceService(DEVICE_SERVICE_APU_TRANSFER, this.scheduledDeadline);
		}
	}

	public completeService(): void {
		const completedDeadline = this.scheduledDeadline;
		this.scheduler.cancelDeviceService(DEVICE_SERVICE_APU_TRANSFER);
		this.completeBatch();
		this.scheduleBatch(completedDeadline);
	}

	public writeAddress(word: number): void {
		this.currentAddress = word & (APU_SAMPLE_RAM_ADDRESS_MASK & ~3);
	}

	public readCpuData(): number {
		return this.dataLatch;
	}

	public writeCpuData(word: number): void {
		this.dataLatch = word >>> 0;
		if (this.mode === APU_TRANSFER_MODE_MANUAL_WRITE) {
			this.sampleMemory.writeWord(this.currentAddress, this.dataLatch);
			this.currentAddress = (this.currentAddress + 4) & APU_SAMPLE_RAM_ADDRESS_MASK;
		}
	}

	public readDmaData(grantEnd: boolean): number {
		if (this.mode === APU_TRANSFER_MODE_DMA_READ && this.fifoCount !== 0) {
			this.dataLatch = this.popFifo();
			this.memory.writeIoValue(IO_APU_TRANSFER_DATA, this.dataLatch);
			this.updateDmaRequests();
			if (grantEnd) {
				this.scheduleBatch(this.scheduler.currentNowCycles());
			}
		}
		return this.dataLatch;
	}

	public writeDmaData(word: number, grantEnd: boolean): void {
		this.dataLatch = word >>> 0;
		if (this.mode === APU_TRANSFER_MODE_DMA_WRITE) {
			this.pushFifo(this.dataLatch);
			this.updateDmaRequests();
			if (grantEnd) {
				this.scheduleBatch(this.scheduler.currentNowCycles());
			}
		}
	}

	public writeControl(word: number): void {
		const mode = word & APU_TRANSFER_MODE_MASK;
		if (mode !== this.mode) {
			this.cancelBatch();
			this.clearFifo();
			this.timingCarry = 0;
			this.mode = mode;
		}
		this.updateDmaRequests();
		this.scheduleBatch(this.scheduler.currentNowCycles());
	}

	private clearFifo(): void {
		this.fifoReadIndex = 0;
		this.fifoWriteIndex = 0;
		this.fifoCount = 0;
	}

	private completeBatch(): void {
		const transferWords = this.scheduledWords;
		this.scheduledWords = 0;
		if (this.mode === APU_TRANSFER_MODE_DMA_READ) {
			for (let index = 0; index < transferWords; index += 1) {
				this.pushFifo(this.sampleMemory.readWord(this.currentAddress));
				this.currentAddress = (this.currentAddress + 4) & APU_SAMPLE_RAM_ADDRESS_MASK;
			}
		} else {
			for (let index = 0; index < transferWords; index += 1) {
				this.sampleMemory.writeWord(this.currentAddress, this.popFifo());
				this.currentAddress = (this.currentAddress + 4) & APU_SAMPLE_RAM_ADDRESS_MASK;
			}
		}
		this.updateDmaRequests();
	}

	private scheduleBatch(anchorCycle: number): void {
		if (this.scheduledWords !== 0) {
			return;
		}
		let words = 0;
		if (this.mode === APU_TRANSFER_MODE_DMA_WRITE) {
			words = this.fifoCount;
		} else if (this.mode === APU_TRANSFER_MODE_DMA_READ) {
			words = APU_TRANSFER_FIFO_WORD_CAPACITY - this.fifoCount;
		}
		if (words === 0) {
			this.scheduledDeadline = 0;
			return;
		}
		const cycles = cyclesUntilBudgetUnits(this.cpuHz, APU_TRANSFER_WORDS_PER_SECOND, this.timingCarry, words);
		this.timingCarry = (APU_TRANSFER_WORDS_PER_SECOND * cycles + this.timingCarry) % this.cpuHz;
		this.scheduledWords = words;
		this.scheduledDeadline = anchorCycle + cycles;
		this.scheduler.scheduleDeviceService(DEVICE_SERVICE_APU_TRANSFER, this.scheduledDeadline);
	}

	private cancelBatch(): void {
		this.scheduler.cancelDeviceService(DEVICE_SERVICE_APU_TRANSFER);
		this.scheduledWords = 0;
		this.scheduledDeadline = 0;
	}

	private pushFifo(word: number): void {
		if (this.fifoCount === APU_TRANSFER_FIFO_WORD_CAPACITY) {
			return;
		}
		this.fifoWords[this.fifoWriteIndex] = word;
		this.fifoWriteIndex = (this.fifoWriteIndex + 1) & (APU_TRANSFER_FIFO_WORD_CAPACITY - 1);
		this.fifoCount += 1;
	}

	private popFifo(): number {
		const word = this.fifoWords[this.fifoReadIndex]!;
		this.fifoReadIndex = (this.fifoReadIndex + 1) & (APU_TRANSFER_FIFO_WORD_CAPACITY - 1);
		this.fifoCount -= 1;
		return word;
	}

	private updateDmaRequests(): void {
		this.dma.setApuDmaWriteReady(this.mode === APU_TRANSFER_MODE_DMA_WRITE && this.fifoCount === 0);
		this.dma.setApuDmaReadReady(this.mode === APU_TRANSFER_MODE_DMA_READ && this.fifoCount === APU_TRANSFER_FIFO_WORD_CAPACITY);
	}
}
